### PR TITLE
Add tracing abstraction, and opencensus impl

### PR DESCRIPTION
**Pull request by AlCutter**
Adds a very simple tracing abstraction, an opencensus implementation, and switches to using it.

### Checklist

- [ ] I have updated the [CHANGELOG](CHANGELOG.md).
  - Adjust the draft version number according to [semantic versioning](https://semver.org/) rules.
- [ ] I have updated [documentation](docs/) accordingly.

### CODE REVIEW

File: monitoring/trace.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+// limitations under the License.
+
+package monitoring
+
+import (
+	"context"
+	"sync"
+)
+
+// StartSpan is the global entry point for Trillian code to create new tracing spans.
```
Comment (daviddrysdale) [inline monitoring/trace.go]: Comment seems to be in the wrong place

File: monitoring/trace.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+// SetStartSpanFunc allows the tracing span implementation to be set.
+// This function will set the global tracing function to the one supplied by
+// the first caller, future calls to this function will have no effect.
+func SetStartSpanFunc(s StartSpanFunc) {
+	once.Do(func() {
+		startSpan = s
+	})
+}
+
+// StartSpan starts a new tracing span.
```
Comment (daviddrysdale) [inline monitoring/trace.go]: Could do with some more explanation of this, particularly the return values (copying the comment from trace.go would work)

File: monitoring/trace.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+// StartSpan is the global entry point for Trillian code to create new tracing spans.
+var (
+	once      sync.Once
+	startSpan StartSpanFunc = func(ctx context.Context, _ string) (context.Context, func()) { return ctx, func() {} }
+)
+
+// SetStartSpanFunc allows the tracing span implementation to be set.
+// This function will set the global tracing function to the one supplied by
+// the first caller, future calls to this function will have no effect.
```
Comment (daviddrysdale) [inline monitoring/trace.go]: Do you actually need the `sync.Once` protection?  Could this whole file be simplified down to just:
```golang
// StartSpan starts a new tracing span using the given message.
// The returned context should be used for all child calls within the span, and
// the returned func should be called to close the span.
// The default implementation of this method is a no-op; insert a real tracing span
// implementation by setting this global variable to the relevant function at start of day.
var StartSpan = func(ctx context.Context, msg string) (context.Context, func()) { return ctx, func() {} }
```

File: server/log_rpc_server.go [hunk 4]
```diff
@@ -811,8 +810,9 @@ func (t *TrillianLogRPCServer) recordIndexPercent(leafIndex int64, treeSize uint
 	}
 }
 
-func spanFor(ctx context.Context, name string) (context.Context, *trace.Span) {
-	return trace.StartSpan(ctx, fmt.Sprintf("%s.%s", traceSpanRoot, name))
+func spanFor(ctx context.Context, name string) (context.Context, func()) {
+	ctx, spanEnd := monitoring.StartSpan(ctx, fmt.Sprintf("%s.%s", traceSpanRoot, name))
+	return ctx, spanEnd
```
Comment (daviddrysdale) [inline server/log_rpc_server.go]: fold these two lines together?

File: monitoring/trace.go (hunk 1 above)
Comment (AlCutter) [reply monitoring/trace.go]: ah, whoops - dinged.

File: monitoring/trace.go (hunk 2 above)
Comment (AlCutter) [reply monitoring/trace.go]: done

File: monitoring/trace.go (hunk 3 above)
Comment (AlCutter) [reply monitoring/trace.go]: Yeah, that's a fair question - I was _slightly_ worried that different implementations might be set in some environments, but it seems quite unlikely.

Changed.

File: server/log_rpc_server.go (hunk 4 above)
Comment (AlCutter) [reply server/log_rpc_server.go]: done

File: monitoring/trace.go [hunk 5]
```diff
@@ -16,28 +16,14 @@ package monitoring
...
-}
+// StartSpan starts a new tracing span using the given message.
+// The returned context should be used for all child calls within the span, and
+// the returned func should be called to close the span.
+// The default implementation of this method is a no-op; insert a real tracing span
+// implementation by setting this global variable to the relevant function at start of day.
+var StartSpan = func(ctx context.Context, msg string) (context.Context, func()) { return ctx, func() {} }
 
 // StartSpanFunc is the signature of a function which starts new tracing spans.
 type StartSpanFunc func(ctx context.Context, name string) (context.Context, func())
```
Comment (daviddrysdale) [inline monitoring/trace.go]: Type no longer used?

Comment (AlCutter) [reply monitoring/trace.go]: type no longer present

**Comment by Martin2112** [comment]
Looks like I'm late to this party.

**Pull request merged**